content/blog/eng/from-0-to-infinite-the-final-keyword.md

```diff
@@ -0,0 +1,338 @@
+
+When we meet a class with the keyword final, who designed the class not allow us inherit from it, if we want to extend some behavior we must do it by composition or other mechanisms (like events, or plugins). 
+
+I think that this is a good method. In fact, in relation to the time used for the analysis of the design class, through final we provide important information on how she should be useds, simplifying the effort of the developer.
+
+# S.O.L.I.D, Liskov and final
+
+Develop architecture isn't a trivial task, many programmers use inheritance as solution to all problems.  
+
+There are many principles and guidelines to help to develop good design 
+like S.O.L.I.D (https://en.wikipedia.org/wiki/SOLID), design patterns, (https://en.wikipedia.org/wiki/Design_pattern) etc.
+But the problem is that these principles are difficult to apply, they almost always lead to writing more code. 
+Not always this is possible, not always this is accepted. 
+
+Regarding final, in my opinion, one of the principle that best represents it is LSP.
+
+Let's see what **Liskov principle** says in more detail. 
+The canonical statement says the following:
+
+> "you should always be able to substitute the parent class with its derived class without any undesirable behaviour"
+
+> The concept of this principle was introduced by Barbara **Liskov** in a 1987 conference keynote and later published in a paper together with Jannette Wing in 1994. 
+
+The more precise definition:
+>“if S is a subtype of T, then objects of type T in a program may be replaced with objects of type S without altering any of >the desirable properties of that program.”
+
+Link to orginal paper: https://dl.acm.org/doi/pdf/10.1145/62138.62141
+
+> Simplifying the following could be good guidelines:
+
+> - Don’t stricter more validation rules on input parameters than those implemented by the parent class.
+> - Objects should be replaceable with instances of their subtypes without altering the correctness of that program
+> - Not add restriction in an overridden method if that restriction doesn't exist in baseline implementation
+
+I have used it like as example because is specifically for the design of hierarchies.
+
+I do not want to write an article on how to respect the development according to **Liskov** in the net there is a lot of literature. I would just like to emphasize that building a hierarchy of objects is not easy, that more the hierarchy is deep then more we are faced with design problems and **final** helps use to prevent this. 
+
+For this reason, before creating subclasses we should ask ourselves questions such as these or similar:
+
+- is my derived class of the same type as the base class?
+- I can use the derived class of base where requireds a base class without having "strange" behaviors in the code at runtime?
```

Comment:
Can I use the **_derived class of base_** (it sounds strange to me, I can't find any similar occurrence on the web)